components/org.wso2.carbon.identity.application.authenticator.google/src/main/java/org/wso2/carbon/identity/application/authenticator/google/GoogleOAuth2Authenticator.java

```diff
@@ -51,13 +63,117 @@ public class GoogleOAuth2Authenticator extends OpenIDConnectAuthenticator {
                 .GOOGLE_TOKEN_ENDPOINT);
         if (StringUtils.isBlank(this.tokenEndpoint)) {
             this.tokenEndpoint = IdentityApplicationConstants.GOOGLE_TOKEN_URL;
         }
     }
 
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+
+        // Google one tap flow does not require any special parameter validation at this level.
+        if (isOneTapEnabled(request)) {
+            return true;
+        }
+        return super.canHandle(request);
+    }
+
+    @Override
+    protected String mapIdToken(AuthenticationContext context, HttpServletRequest request,
+                                OAuthClientResponse oAuthResponse) throws AuthenticationFailedException {
+
+        if (isOneTapEnabled(request)) {
+            Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
+            String clientID = authenticatorProperties.get(OIDCAuthenticatorConstants.CLIENT_ID);
+
+            validateCSRF(request, clientID);
+            boolean validJWT = Utils.validateJWT(request.getParameter(CREDENTIAL), clientID);
+            if (!validJWT) {
+                throw new AuthenticationFailedException(GoogleErrorConstants.ErrorMessages
+                        .TOKEN_VALIDATION_FAILED_ERROR.getCode(), String.format(GoogleErrorConstants.ErrorMessages
+                        .TOKEN_VALIDATION_FAILED_ERROR.getMessage(), clientID));
+            }
+            String idToken = request.getParameter(CREDENTIAL);
+            context.setProperty(OIDCAuthenticatorConstants.ID_TOKEN, idToken);
+            return idToken;
+        }
+        return super.mapIdToken(context, request, oAuthResponse);
+    }
+
+    /**
+     * Validate CSRF based on configuration value for Google One Tap.
+     * Google One Tap UI appears on accounts.asg.io domain so the CSRF cookie comes under accounts.asg.io
+     * Authenticated response comes to api.asg.io domain with CSRF parameter hence CSRF cookie is blocked.
+     * Still there is no configuration from Google side to overcome this issue.
+     * Raising this issue under <a href="https://github.com/wso2/product-is/issues/14779">...</a>
+     * Once this issue is fixed, this configuration check should be removed.
+     *
+     * @param request  Authenticated request for commonauth coming from Google.
+     * @param clientID Google client ID.
+     * @throws AuthenticationFailedException Error when CSRF validation failed.
+     */
+    private void validateCSRF(HttpServletRequest request, String clientID) throws AuthenticationFailedException {
+
+        boolean validateCSRF = true;
+        String enableCSRFValidationForGOT = getAuthenticatorConfig().getParameterMap()
+                .get(GoogleOAuth2AuthenticationConstant.ENABLE_CSRF_VALIDATION_FOR_GOT);
+
+        if (StringUtils.isNotBlank(enableCSRFValidationForGOT)) {
+            validateCSRF = Boolean.parseBoolean(enableCSRFValidationForGOT);
+        }
```

Comment:
Check if this configuration is needed to be added to the master.

Reply:
This will be removed on major release. Keeping until all level testing are over.